File: Week2/StacksAndQueues/ResizingArrayStackOfStringsProject/ResizingArrayStackOfStrings.cpp
/*
 * ResizingArrayStackOfStrings.cpp
 *
 *  Created on: 13 Sep 2018
 *      Author: danielcumberbatch
 */

#include "ResizingArrayStackOfStrings.h"

#include <algorithm>
#include <stdexcept>
#include <iostream>

ResizingArrayStackOfStrings::ResizingArrayStackOfStrings()
  : m_a(new std::string[1])
  , m_capacity(1)
  , m_head(0)
{}

ResizingArrayStackOfStrings::ResizingArrayStackOfStrings(const ResizingArrayStackOfStrings& other)
  : m_a(new std::string[other.m_capacity])
  , m_capacity(other.m_capacity)
  , m_head(other.m_head)
{
  std::copy(other.m_a, other.m_a+other.m_capacity, this->m_a);
}

ResizingArrayStackOfStrings& ResizingArrayStackOfStrings::operator=(const ResizingArrayStackOfStrings& other)
{
  if (this != &other)
    {
      ResizingArrayStackOfStrings tmp(other);
      std::string *p = tmp.m_a;
      tmp.m_a = this->m_a;
      this->m_a = p;

      this->m_capacity = tmp.m_capacity;
      this->m_head = tmp.m_head;
    }

  return *this;
}

ResizingArrayStackOfStrings::~ResizingArrayStackOfStrings()
{
  delete [] this->m_a;
}

bool ResizingArrayStackOfStrings::empty() const
{
  return this->size() == 0;
}

std::size_t ResizingArrayStackOfStrings::size() const
{
  return this->m_head;
}

const std::string& ResizingArrayStackOfStrings::top() const
{
  if (this->empty())
    {
      throw std::logic_error("top: Stack is empty.");
    }

  return this->m_a[this->m_head-1];
}

void ResizingArrayStackOfStrings::push(const std::string& s)
{
  if (this->size() == this->m_capacity)
    {
      const std::size_t new_capacity = std::max<std::size_t>(2*this->size(),1);
      std::string *p = new std::string[new_capacity];
      std::copy(this->m_a, this->m_a+this->m_capacity, p);
      delete [] this->m_a;
      this->m_a = p;

      this->m_capacity = new_capacity;
    }

  this->m_a[this->m_head++] = s;
}

void ResizingArrayStackOfStrings::pop()
{
  if (this->empty())
    {
      throw std::logic_error("pop: Stack is empty.");
    }

  --this->m_head;

  if (this->m_head <= this->m_capacity/4)
    {
      const std::size_t new_capacity = std::max<std::size_t>(1,this->m_capacity/4);
      std::string *p = new std::string[new_capacity];
      std::copy(this->m_a, this->m_a+this->m_head, p);
      delete [] this->m_a;
      this->m_a = p;

      this->m_capacity = new_capacity;
    }
}

std::size_t ResizingArrayStackOfStrings::capacity() const
{
  return this->m_capacity;
}

File: Week2/StacksAndQueues/StacksAndQueues/Java/ResizingArrayStackOfStrings/src/ResizingArrayStackOfStrings.java

public class ResizingArrayStackOfStrings {
	private String[] a;
	private int capacity;
	private int head;
	
	public ResizingArrayStackOfStrings() {
		this.a = new String[1];
		this.capacity = this.a.length;
		this.head = 0;
	}
	
	public boolean isEmpty() {
		return this.head == 0;
	}
	
	public String pop() {
		if (this.isEmpty()) {
			throw new IllegalStateException("Cannot pop empty ResizingArrayStackOfStrings.");
		}
		
		--this.head;
		final String s = this.a[this.head];
		this.a[this.head] = null;
		
		if (this.head == this.capacity/4) {
			final int newCapacity = Math.max(1, this.capacity/2);
			final String[] newA = new String[newCapacity];
			for (int i = 0; i < newCapacity; ++i) {
				newA[i] = this.a[i];
			}

			this.a = newA;
			this.capacity = newCapacity;
		}
		
		return s;
	}
	
	public void push(String s) {
		if (this.head == this.capacity) {
			final int newCapacity = Math.max(1, 2*this.capacity);
			final String[] newA = new String[newCapacity];
			for (int i = 0; i < this.capacity; ++i) {
				newA[i] = this.a[i];
			}
			
			this.capacity = newCapacity;
			this.a = newA;
		}
		
		this.a[head++] = s;
	}
	
	public int size() {
		return this.head;
	}
}
